float32[] times
float64[] states
float32 horizon